File: nodejs/itemlocking/src/item-locking.js
const {CacheManager,CacheItem,LockHandle,TimeSpan, Expiration,JsonDataType, ExpirationType} = require('ncache-client');
let cache = null;
const config = require('./app.config.json');
async function Run() {
    // Initialize cache 
    await InitializeCache();
    await cache.clear();
    // Create a simple customer object
    let customer = CreateNewCustomer();
    const key = GetKey(customer);

    // Adding item
    await AddObjectToCache(key, customer);

    const lockHandle = new LockHandle();
    const timeSpan = new TimeSpan(0,0,0,20);
    // Get the object from cache
    await GetObjectFromCache(key,timeSpan,lockHandle);

    // Lock Item in cache 
    await LockItemInCache(key, timeSpan,lockHandle);

    // Unlock item in cache using multiple ways
    await UnLockItemInCache(key, lockHandle);

    // Delete the existing object
    await RemoveObjectFromCache(key);

    //close cache
    await cache.close();

}

async function InitializeCache() {
    // Get Cache Name
    const cacheName = config.cacheProperties.cacheName
    if (!cacheName) {
        console.log("The CacheID cannot be null or empty.");
        return;
    }
    // Initialize an instance of the cache to begin performing operations:
    cache = await CacheManager.getCache(cacheName);

    // Print output on console
    console.log(`Cache ${cacheName} is initialized.`);
            
}

async function AddObjectToCache(key, customer) {
    // creating time span for setting expiration interval
    const expirationInterval = new TimeSpan(0, 1, 0);

    // creating expiration interval with absolute type to assign to cache item
    const expiration =  new Expiration(ExpirationType.Absolute);
    expiration.setExpireAfter(expirationInterval)

    //Populating cache item and adding new expiration
    const item = new CacheItem(customer,"Customer");
    item.Expiration = expiration;

    // Adding cacheitem to cache with an absolute expiration of 1 minute
    await cache.add(key, item);

    // Print output on console
    console.log("\nObject is added to cache.");
}

async function GetObjectFromCache(key,timeSpan,lockHandle) {
    const cachedCustomer = await cache.get(key,JsonDataType.Object,null,null,true,timeSpan,lockHandle);
    PrintCustomerDetails(cachedCustomer);
}

async function RemoveObjectFromCache(key) {
    // Remove the existing customer
    await cache.remove(key,JsonDataType.Object);

    // Print output on console
    console.log("\nObject is removed from cache.");
}
async function LockItemInCache(key, timeSpan, lockHandle) {
    const isLocked = await cache.lock(key,timeSpan,lockHandle);
    if (!isLocked)
    {
        console.log("Lock acquired on " + lockHandle.getLockId());
    }
}
async function UnLockItemInCache(key, lockHandle) {
    await cache.unlock(key,lockHandle)
}
function GetKey(customer) {
    return `Customer:${customer.CustomerID}`;
}
function CreateNewCustomer() {
    return({
        CustomerID : "DAVJO",
        ContactName : "David Johnes",
        CompanyName : "Lonesome Pine Restaurant",
        ContactNo : "12345-6789",
        Address : "Silicon Valley, Santa Clara, California",
    });
}
function PrintCustomerDetails(customer) {
    if (customer == null) return;

        console.log();
        console.log("Customer Details are as follows: ");
        console.log("ContactName: " + customer.ContactName);
        console.log("CompanyName: " + customer.CompanyName);
        console.log("Contact No: " + customer.ContactNo);
        console.log("Address: " + customer.Address);
        console.log();

}
module.exports = {
    Run
}

File: python/item_locking/item_locking.py
from ncache.client.CacheManager import CacheManager
from ncache.client.CacheItem import CacheItem
from ncache.client.LockHandle import LockHandle
from ncache.runtime.util.TimeSpan import TimeSpan
from sample_data.product import Product


def initialize_cache():
    cache_name = "demoCache"
    cache = CacheManager.get_cache(cache_name)
    print("\nCache initialized successfully.")
    return cache


def create_product():
    product = Product(id="1001", name="Tea", unit_price=50, units_available=20)
    return product


def get_key(product):
    key = "Product:" + product.id
    return key


def add_item_in_cache(cache, key, product):
    item = CacheItem(product)
    cache.add(key, item)
    print("\nObject is added in cache.")


def get_item_from_cache(cache, key, lock_handle, time_span):
    fetched_product = cache.get(key, Product, None, None, True, time_span, lock_handle)
    print("\nObject is fetched from cache.")
    print(fetched_product)
    print("\nLock acquired on: ", lock_handle.get_lock_id())


def lock_item_in_cache(cache, key, lock_handle, time_span):
    is_locked = cache.lock(key, time_span, lock_handle)
    if not is_locked:
        print("\nLock acquired on: ", lock_handle.get_lock_id())


def unlock_item_in_cache(key, lock_handle):
    cache.unlock(key, lock_handle)
    print("\nThe item is unlocked.")


def remove_item_from_cache(cache, key):
    # Count before removal
    count = cache.get_count()
    print("\nNumber of items in cache before removal: ", count)
    cache.remove(key, Product)
    print("\nObject is removed from cache.")
    # Count after removal
    count = cache.get_count()
    print("\nNumber of items in cache after removal: ", count)


def dispose_cache(cache):
    cache.close()
    print("\nCache closed.")


# Initialize instance of the cache to perform operations
cache = initialize_cache()

# Create a Product object
product = create_product()
key = get_key(product)

# Add item in cache
add_item_in_cache(cache, key, product)

# Create new lock handle to fetch item using locking
lock_handle = LockHandle()

# Time span for which lock will be taken
time_span = TimeSpan(ticks=0, hours=0, minutes=0, seconds=20)

# Get item from cache
get_item_from_cache(cache, key, lock_handle, time_span)

# Lock item in cache
lock_item_in_cache(cache, key, lock_handle, time_span)

# Unlock item in cache
unlock_item_in_cache(key, lock_handle)

# Remove item from cache
remove_item_from_cache(cache, key)

# Dispose the cache once done
dispose_cache(cache)
